runtime/jailer.go

```diff
@@ -0,0 +1,303 @@
+
+	"github.com/firecracker-microvm/firecracker-go-sdk"
+)
+
+const (
+	jailerFolder          = "jail"
+	firecrackerBinName    = "firecracker"
+	kernelImageFileName   = "kernel-image"
+	jailerHandlerName     = "firecracker-containerd-jail-handler"
+	jailerFifoHandlerName = "firecracker-containerd-jail-fifo-handler"
+	runcConfigPath        = "/etc/containerd/firecracker-runc-config.json"
+
+	// JailingOn is used to signify whether or not jailing has been turned on
+	JailingOn = "on"
+	// JailingOff is used to signify that the jailer should not be used
+	JailingOff = "off"
+)
+
+// exposeFileToJail will inspect the given file, srcFilePath, and based on the
+// file type, proper handling will occur to ensure that the file is visible in
+// the jail.
+func (j jailer) exposeFileToJail(srcFilePath string, uid, gid int) error {
+	stat := syscall.Stat_t{}
+	if err := syscall.Stat(srcFilePath, &stat); err != nil {
+		return err
+	}
+
+	path := j.RootPath()
+	dst := ""
+	if (stat.Mode & syscall.S_IFMT) == syscall.S_IFBLK { // TODO: Fix this
+		path = filepath.Join(path, "dev")
+		dst = filepath.Join(path, filepath.Base(srcFilePath))
+		if err := exposeBlockDeviceToJail(dst, int(stat.Rdev)); err != nil {
+			return err
+		}
+	} else {
+		return fmt.Errorf("unsupported mode: %v", stat.Mode)
+	}
+
+	if err := os.Chmod(dst, 0600); err != nil {
+		return err
+	}
+
+	if err := os.Chown(dst, uid, gid); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// exposeBlockDeviceToJail will call mknod on the block device to ensure
+// visibility of the device
+func exposeBlockDeviceToJail(dst string, rdev int) error {
+	if err := syscall.Mknod(dst, syscall.S_IFBLK, int(rdev)); err != nil {
+		return err
+	}
+	return nil
+}
+
+type jailer struct {
+	jailPath string
+}
+
+func newJailer(rootpath string) jailer {
+	return jailer{
+		jailPath: rootpath,
+	}
+}
+
+// JailPath returns the base directory from where the jail binary will be ran
+// from
+func (j jailer) JailPath() string {
+	return j.jailPath
+}
+
+// RootPath returns the root fs of the jailed system.
+func (j jailer) RootPath() string {
+	return filepath.Join(j.jailPath, "rootfs")
+}
+
+func (j jailer) ContentsPath() string {
+	return filepath.Join(j.RootPath(), "var", "lib", "firecracker-containerd")
+}
+
+// BuildHandler will link the necessary files except for the fifos due to the
+// fifos needing to be created. Also, this will create the proper device nodes
+// needed by Firecracker
+func (j *jailer) BuildHandler(logger *logrus.Entry, cfg *Config, socketPath *string, vmID string) firecracker.Handler {
+	jailPath := j.JailPath()
+	rootPath := j.RootPath()
+	*socketPath = filepath.Join(rootPath, "api.socket")
+
+	return firecracker.Handler{
+		Name: jailerHandlerName,
+		Fn: func(ctx context.Context, m *firecracker.Machine) error {
+			// Create the proper paths needed for the runc jailer
+			logger.Debugf("Creating root drive path %v", rootPath)
+			if err := os.MkdirAll(rootPath, 0777); err != nil {
```

Comment:
How about using 0700? A bit scary to see the directory is writable for everyone.

Reply:
Done